jdk/src/bsd/doc/man/apt.1
." Copyright (c) 2004, 2012, Oracle and/or its affiliates. All rights reserved.
." DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
."
." This code is free software; you can redistribute it and/or modify it
." under the terms of the GNU General Public License version 2 only, as
." published by the Free Software Foundation.
."
." This code is distributed in the hope that it will be useful, but WITHOUT
." ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
." FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
." version 2 for more details (a copy is included in the LICENSE file that
." accompanied this code).
."
." You should have received a copy of the GNU General Public License version
." 2 along with this work; if not, write to the Free Software Foundation,
." Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
."
." Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
." or visit www.oracle.com if you need additional information or have any
." questions.
."
.TH apt 1 "10 May 2011"

.LP
.SH "NAME"
.LP
.LP
\f2apt\fP \- annotation processing tool
.LP
.SH "SYNOPSIS"
.LP
.LP
\f2apt [\-classpath \fP\f2classpath\fP] [\-sourcepath \f2sourcepath\fP] [\-d \f2directory\fP] [\-s \f2directory\fP] [\-factorypath \f2path\fP] [\-factory \f2class\fP] [\-print] [\-nocompile] [\-A\f2key\fP[\f2=val\fP] ...] [\f2javac option\fP] sourcefiles [@files]
.LP
.SH "PARAMETERS"
.LP
.LP
Options may be in any order. For a discussion of parameters which apply to a specific option, see OPTIONS below.
.LP
.RS 3
.TP 3
sourcefiles
Zero or more source files to be processed.
.TP 3
@files
One or more files that list source files or other options
.RE

.LP
.SH "DESCRIPTION"
.LP
.LP
\f3Note\fP: The \f2apt\fP tool and its associated API contained in the package \f2com.sun.mirror\fP have been deprecated since JDK 7 and are planned to be removed in the next major JDK release. Use the options available in the \f2javac(1)\fP tool and the APIs contained in the packages \f2javax.annotation.processing\fP and \f2javax.lang.model\fP to process annotations.
.LP
.LP
The tool \f2apt\fP, annotation processing tool, includes reflective APIs and supporting infrastructure to process program annotations. The \f2apt\fP reflective APIs provide a build\-time, source\-based, read\-only view of program structure. These reflective APIs are designed to cleanly model the Java(TM) programming language's type system after the addition of generics. First, \f2apt\fP runs annotation processors that can produce new source code and other files. Next, \f2apt\fP can cause compilation of both original and generated source files, easing development. The reflective APIs and other APIs used to interact with the tool are subpackages of \f2com.sun.mirror\fP.
.LP
.LP
A fuller discussion of how the tool operates as well as instructions for developing with \f2apt\fP are in
.na
\f4Getting Started with \fP\f4apt\fP. @
.fi
http://download.oracle.com/javase/7/docs/technotes/guides/apt/GettingStarted.html
.LP
.SH "OPTIONS"
.LP
.SS
apt specific options
.LP
.RS 3
.TP 3
\-s dir
Specify the directory root under which processor\-generated source files will be placed; files are placed in subdirectories based on package namespace.
.TP 3
\-nocompile
Do not compile source files to class files.
.TP 3
\-print
Print out textual representation of specified types; perform no annotation processing or compilation.
.TP 3
\-A[key[=val]]
Options to pass to annotation processors \-\- these are not interpreted by \f2apt\fP directly, but are made available for use by individual processors
.TP 3
\-factorypath path
Specify where to find annotation processor factories; if this option is used, the classpath is \f2not\fP searched for factories.
.TP 3
\-factory classname
Name of annotation processor factory to use; bypasses default discovery process
.TP 3
\-version
Print version information.
.TP 3
\-X
Display information about non\-standard options.
.RE

.LP
.SS
Options shared with javac
.LP
.RS 3
.TP 3
\-d dir
Specify where to place processor and javac generated class files
.TP 3
\-cp path or \-classpath path
Specify where to find user class files and annotation processor factories. If \f2\-factorypath\fP is given, the classpath is not searched for factories.
.RE

.LP
.LP
Consult the javac(1) man page for information on \f2javac\fP options.
.LP
.SS
Non\-Standard Options
.LP
.RS 3
.TP 3
\-XListAnnotationTypes
List found annotation types.
.TP 3
\-XListDeclarations
List specified and included declarations.
.TP 3
\-XPrintAptRounds
Print information about initial and recursive \f2apt\fP rounds.
.TP 3
\-XPrintFactoryInfo
Print information about which annotations a factory is asked to process.
.TP 3
\-XclassesAsDecls
Treat both class and source files as declarations to process.
.RE

.LP
.LP
\f3Note\fP: Because these options are non\-standard, they are subject to change without notice.
.LP
.SH "NOTES"
.LP
.LP
The \f2apt\fP tool and its associated API contained in the package \f2com.sun.mirror\fP have been deprecated since JDK 7 and are planned to be removed in the next major JDK release. Use the options available in the \f2javac(1)\fP tool and the APIs contained in the packages \f2javax.annotation.processing\fP and \f2javax.lang.model\fP to process annotations.
.LP
.SH "SEE ALSO"
.LP
.RS 3
.TP 2
o
javac(1), java(1)
.RE

.LP



langtools/src/share/classes/com/sun/tools/apt/comp/Apt.java
/*
 * Copyright (c) 2004, 2011, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package com.sun.tools.apt.comp;

import com.sun.tools.javac.code.*;
import com.sun.tools.javac.comp.*;
import com.sun.tools.javac.tree.*;
import com.sun.tools.javac.util.*;
import com.sun.tools.javac.tree.TreeScanner;
import com.sun.tools.javac.util.Context;
import com.sun.tools.apt.util.Bark;
import com.sun.tools.javac.util.Position;

import java.util.*;
import java.util.regex.*;
import java.lang.reflect.*;
import java.lang.reflect.InvocationTargetException;
import java.io.IOException;

import com.sun.tools.apt.*;
import com.sun.tools.apt.comp.*;
import com.sun.tools.javac.code.Symbol.*;

import com.sun.mirror.declaration.TypeDeclaration;
import com.sun.mirror.declaration.AnnotationTypeDeclaration;
import com.sun.mirror.apt.*;
// import com.sun.mirror.apt.AnnotationProcessorFactory;
import com.sun.mirror.apt.AnnotationProcessors;

import com.sun.tools.apt.mirror.AptEnv;
import com.sun.tools.apt.mirror.apt.FilerImpl;
import com.sun.tools.apt.mirror.apt.AnnotationProcessorEnvironmentImpl;


import static com.sun.tools.apt.mirror.declaration.DeclarationMaker.isJavaIdentifier;

/**
 * Apt compiler phase.
 *
 *  <p><b>This is NOT part of any supported API.
 *  If you write code that depends on this, you do so at your own
 *  risk.  This code and its internal interfaces are subject to change
 *  or deletion without notice.</b>
 */
@SuppressWarnings("deprecation")
public class Apt extends ListBuffer<Env<AttrContext>> {
    java.util.Set<String> genSourceFileNames = new java.util.LinkedHashSet<String>();
    public java.util.Set<String> getSourceFileNames() {
        return genSourceFileNames;
    }

    /** List of names of generated class files.
     */
    java.util.Set<String> genClassFileNames  = new java.util.LinkedHashSet<String>();
    public java.util.Set<String> getClassFileNames() {
        return genClassFileNames;
    }

    /* AptEnvironment */
    AptEnv aptenv;

    private Context context;

    /** The context key for the todo list. */

    protected static final Context.Key<Apt> aptKey =
        new Context.Key<Apt>();

    /** Get the Apt instance for this context. */
    public static Apt instance(Context context) {
        Apt instance = context.get(aptKey);
        if (instance == null)
            instance = new Apt(context);
        return instance;
    }

    /** Create a new apt list. */
    protected Apt(Context context) {
        this.context = context;

        context.put(aptKey, this);
        aptenv = AptEnv.instance(context);
    }

    /**
     * Used to scan javac trees to build data structures needed for
     * bootstrapping the apt environment.  In particular:
     *
     * <ul>
     *
     * <li> Generate list of canonical names of annotation types that
     * appear in source files given on the command line
     *
     * <li> Collect list of javac symbols representing source files
     * given on the command line
     *
     * </ul>
     */
    static class AptTreeScanner extends TreeScanner {

        // Set of fully qualified names of annotation types present in
        // examined source
        private Set<String> annotationSet;

        // Symbols to build bootstrapping declaration list
        private Collection<ClassSymbol> specifiedDeclCollection;
        private Collection<ClassSymbol> declCollection;

        public Set<String> getAnnotationSet() {
            return annotationSet;
        }

        public AptTreeScanner() {
            annotationSet = new  LinkedHashSet<String>();
            specifiedDeclCollection = new LinkedHashSet<ClassSymbol>();
            declCollection = new LinkedHashSet<ClassSymbol>();
        }

        public void visitTopLevel(JCTree.JCCompilationUnit tree) {
            super.visitTopLevel(tree);
            // Print out contents -- what are we dealing with?

            for(JCTree d: tree.defs) {
                if (d instanceof JCTree.JCClassDecl)
                    specifiedDeclCollection.add(((JCTree.JCClassDecl) d).sym);
            }

        }

        public void visitBlock(JCTree.JCBlock tree) {
            ; // Do nothing.
        }


        // should add nested classes to packages, etc.
        public void visitClassDef(JCTree.JCClassDecl tree) {
            if (tree.sym == null) {
                // could be an anon class w/in an initializer
                return;
            }

            super.visitClassDef(tree);

            declCollection.add(tree.sym);
        }

        public void visitMethodDef(JCTree.JCMethodDecl tree) {
            super.visitMethodDef(tree);
        }

        public void visitVarDef(JCTree.JCVariableDecl tree) {
            super.visitVarDef(tree);
        }

        public void visitAnnotation(JCTree.JCAnnotation tree) {
            super.visitAnnotation(tree);
            annotationSet.add(tree.type.tsym.toString());
        }
    }

    Set<String> computeAnnotationSet(Collection<ClassSymbol> classSymbols) {
        Set<String> annotationSet = new HashSet<String>();

        for(ClassSymbol classSymbol: classSymbols) {
            computeAnnotationSet(classSymbol, annotationSet);
        }
        return annotationSet;
    }

    void computeAnnotationSet(Symbol symbol, Set<String> annotationSet) {
        if (symbol != null ) {
            if (symbol.getAnnotationMirrors() != null)
                for(Attribute.Compound compound: symbol.getAnnotationMirrors())
                    annotationSet.add(compound.type.tsym.toString()); // should fullName be used instead of toString?

            if (symbol instanceof Symbol.MethodSymbol) // add parameter annotations
                for(Symbol param: ((MethodSymbol) symbol).params())
                    computeAnnotationSet(param, annotationSet);

            if (symbol.members() != null) {
                for(Scope.Entry e = symbol.members().elems; e != null; e = e.sibling)
                    computeAnnotationSet(e.sym, annotationSet);
            }
        }
    }

    public void main(com.sun.tools.javac.util.List<JCTree.JCCompilationUnit> treeList,
                     ListBuffer<ClassSymbol> classes,
                     Map<String, String> origOptions,
                     ClassLoader aptCL,
                     AnnotationProcessorFactory providedFactory,
                     java.util.Set<Class<? extends AnnotationProcessorFactory> > productiveFactories) {
        Bark bark = Bark.instance(context);
        java.io.PrintWriter out = bark.getWriter(Log.WriterKind.WARNING);
        Options options = Options.instance(context);

        Collection<TypeDeclaration> spectypedecls =     new LinkedHashSet<TypeDeclaration>();
        Collection<TypeDeclaration> typedecls =         new LinkedHashSet<TypeDeclaration>();
        Set<String> unmatchedAnnotations =              new LinkedHashSet<String>();
        Set<AnnotationTypeDeclaration> emptyATDS =      Collections.emptySet();
        Set<Class<? extends AnnotationProcessorFactory> > currentRoundFactories =
            new LinkedHashSet<Class<? extends AnnotationProcessorFactory> >();

        // Determine what annotations are present on the input source
        // files, create collections of specified type declarations,
        // and type declarations.
        AptTreeScanner ats = new AptTreeScanner();
        for(JCTree t: treeList) {
            t.accept(ats);
        }

        // Turn collection of ClassSymbols into Collection of apt decls
        for (ClassSymbol cs : ats.specifiedDeclCollection) {
            TypeDeclaration decl = aptenv.declMaker.getTypeDeclaration(cs);
            spectypedecls.add(decl);
        }

        for (ClassSymbol cs : ats.declCollection) {
            TypeDeclaration decl = aptenv.declMaker.getTypeDeclaration(cs);
            typedecls.add(decl);
        }

        unmatchedAnnotations.addAll(ats.getAnnotationSet());

        // Process input class files
        for(ClassSymbol cs : classes) {
            TypeDeclaration decl = aptenv.declMaker.getTypeDeclaration(cs);
            // System.out.println("Adding a class to spectypedecls");
            spectypedecls.add(decl);
            typedecls.add(decl);
            computeAnnotationSet(cs, unmatchedAnnotations);
        }

        if (options.get("-XListAnnotationTypes") != null) {
            out.println("Set of annotations found:" +
                        (new TreeSet<String>(unmatchedAnnotations)).toString());
        }

        AnnotationProcessorEnvironmentImpl trivAPE =
            new AnnotationProcessorEnvironmentImpl(spectypedecls, typedecls, origOptions, context);

        if (options.get("-XListDeclarations") != null) {
            out.println("Set of Specified Declarations:" +
                        spectypedecls);

            out.println("Set of Included Declarations: " +
                           typedecls);
        }

        if (options.get("-print") != null) {
            if (spectypedecls.size() == 0 )
                throw new UsageMessageNeededException();

            // Run the printing processor
            AnnotationProcessor proc = (new BootstrapAPF()).getProcessorFor(new HashSet<AnnotationTypeDeclaration>(),
                                                                            trivAPE);
            proc.process();
        } else {
            // Discovery process

            // List of annotation processory factory instances
            java.util.Iterator<AnnotationProcessorFactory> providers = null;
            {
                /*
                 * If a factory is provided by the user, the
                 * "-factory" and "-factorypath" options are not used.
                 *
                 * Otherwise, if the "-factory" option is used, search
                 * the appropriate path for the named class.
                 * Otherwise, use sun.misc.Service to implement the
                 * default discovery policy.
                 */

                java.util.List<AnnotationProcessorFactory> list =
                    new LinkedList<AnnotationProcessorFactory>();
                String factoryName = options.get("-factory");

                if (providedFactory != null) {
                    list.add(providedFactory);
                    providers = list.iterator();
                } else if (factoryName != null) {
                    try {
                        AnnotationProcessorFactory factory =
                            (AnnotationProcessorFactory) (aptCL.loadClass(factoryName).newInstance());
                        list.add(factory);
                    } catch (ClassNotFoundException cnfe) {
                        bark.aptWarning("FactoryNotFound", factoryName);
                    } catch (ClassCastException cce) {
                        bark.aptWarning("FactoryWrongType", factoryName);
                    } catch (Exception e ) {
                        bark.aptWarning("FactoryCantInstantiate", factoryName);
                    } catch(Throwable t) {
                        throw new AnnotationProcessingError(t);
                    }

                    providers = list.iterator();
                } else {
                    @SuppressWarnings("unchecked")
                    Iterator<AnnotationProcessorFactory> iter =
                            sun.misc.Service.providers(AnnotationProcessorFactory.class, aptCL);
                    providers = iter;

                }
            }

            java.util.Map<AnnotationProcessorFactory, Set<AnnotationTypeDeclaration>> factoryToAnnotation =
                new LinkedHashMap<AnnotationProcessorFactory, Set<AnnotationTypeDeclaration>>();

            if (!providers.hasNext() && productiveFactories.size() == 0) {
                if (unmatchedAnnotations.size() > 0)
                    bark.aptWarning("NoAnnotationProcessors");
                if (spectypedecls.size() == 0)
                    throw new UsageMessageNeededException();
                return; // no processors; nothing else to do
            } else {
                // If there are no annotations, still give
                // processors that match everything a chance to
                // run.

                if(unmatchedAnnotations.size() == 0)
                    unmatchedAnnotations.add("");

                Set<String> emptyStringSet = new HashSet<String>();
                emptyStringSet.add("");
                emptyStringSet = Collections.unmodifiableSet(emptyStringSet);

                while (providers.hasNext() ) {
                    Object provider = providers.next();
                    try {
                        Set<String> matchedStrings = new HashSet<String>();

                        AnnotationProcessorFactory apf = (AnnotationProcessorFactory) provider;
                        Collection<String> supportedTypes = apf.supportedAnnotationTypes();

                        Collection<Pattern> supportedTypePatterns = new LinkedList<Pattern>();
                        for(String s: supportedTypes)
                            supportedTypePatterns.add(importStringToPattern(s));

                        for(String s: unmatchedAnnotations) {
                            for(Pattern p: supportedTypePatterns) {
                                if (p.matcher(s).matches()) {
                                    matchedStrings.add(s);
                                    break;
                                }
                            }
                        }

                        unmatchedAnnotations.removeAll(matchedStrings);

                        if (options.get("-XPrintFactoryInfo") != null) {
                            out.println("Factory " + apf.getClass().getName() +
                                        " matches " +
                                        ((matchedStrings.size() == 0)?
                                         "nothing.": matchedStrings));
                        }

                        if (matchedStrings.size() > 0) {
                            // convert annotation names to annotation
                            // type decls
                            Set<AnnotationTypeDeclaration> atds = new HashSet<AnnotationTypeDeclaration>();

                            // If a "*" processor is called on the
                            // empty string, pass in an empty set of
                            // annotation type declarations.
                            if (!matchedStrings.equals(emptyStringSet)) {
                                for(String s: matchedStrings) {
                                    TypeDeclaration decl = aptenv.declMaker.getTypeDeclaration(s);
                                    AnnotationTypeDeclaration annotdecl;
                                    if (decl == null) {
                                        bark.aptError("DeclarationCreation", s);
                                    } else {
                                        try {
                                            annotdecl = (AnnotationTypeDeclaration)decl;
                                            atds.add(annotdecl);

                                        } catch (ClassCastException cce) {
                                            bark.aptError("BadDeclaration", s);
                                        }
                                    }
                                }
                            }

                            currentRoundFactories.add(apf.getClass());
                            productiveFactories.add(apf.getClass());
                            factoryToAnnotation.put(apf, atds);
                        } else if (productiveFactories.contains(apf.getClass())) {
                            // If a factory provided a processor in a
                            // previous round but doesn't match any
                            // annotations this round, call it with an
                            // empty set of declarations.
                            currentRoundFactories.add(apf.getClass());
                            factoryToAnnotation.put(apf, emptyATDS );
                        }

                        if (unmatchedAnnotations.size() == 0)
                            break;

                    } catch (ClassCastException cce) {
                        bark.aptWarning("BadFactory", cce);
                    }
                }

                unmatchedAnnotations.remove("");
            }

            // If the set difference of productiveFactories and
            // currentRoundFactories is non-empty, call the remaining
            // productive factories with an empty set of declarations.
            {
                java.util.Set<Class<? extends AnnotationProcessorFactory> > neglectedFactories =
                    new LinkedHashSet<Class<? extends AnnotationProcessorFactory>>(productiveFactories);
                neglectedFactories.removeAll(currentRoundFactories);
                for(Class<? extends AnnotationProcessorFactory> working : neglectedFactories) {
                    try {
                        AnnotationProcessorFactory factory = working.newInstance();
                        factoryToAnnotation.put(factory, emptyATDS);
                    } catch (Exception e ) {
                        bark.aptWarning("FactoryCantInstantiate", working.getName());
                    } catch(Throwable t) {
                        throw new AnnotationProcessingError(t);
                    }
                }
            }

            if (unmatchedAnnotations.size() > 0)
                bark.aptWarning("AnnotationsWithoutProcessors", unmatchedAnnotations);

            Set<AnnotationProcessor> processors = new LinkedHashSet<AnnotationProcessor>();

            // If there were no source files AND no factory matching "*",
            // make sure the usage message is printed
            if (spectypedecls.size() == 0 &&
                factoryToAnnotation.keySet().size() == 0 )
                throw new UsageMessageNeededException();

            try {
                for(Map.Entry<AnnotationProcessorFactory, Set<AnnotationTypeDeclaration>> entry :
                        factoryToAnnotation.entrySet()) {
                    AnnotationProcessorFactory  apFactory = entry.getKey();
                    AnnotationProcessor processor = apFactory.getProcessorFor(entry.getValue(),
                                                                              trivAPE);
                    if (processor != null)
                        processors.add(processor);
                    else
                        bark.aptWarning("NullProcessor", apFactory.getClass().getName());
                }
            } catch(Throwable t) {
                throw new AnnotationProcessingError(t);
            }

            LinkedList<AnnotationProcessor> temp = new LinkedList<AnnotationProcessor>();
            temp.addAll(processors);

            AnnotationProcessor proc = AnnotationProcessors.getCompositeAnnotationProcessor(temp);

            try {
                proc.process();
            } catch (Throwable t) {
                throw new AnnotationProcessingError(t);
            }

            // Invoke listener callback mechanism
            trivAPE.roundComplete();

            FilerImpl filerimpl = (FilerImpl)trivAPE.getFiler();
            genSourceFileNames = filerimpl.getSourceFileNames();
            genClassFileNames = filerimpl.getClassFileNames();
            filerimpl.flush(); // Make sure new files are written out
        }
    }

    /**
     * Convert import-style string to regex matching that string.  If
     * the string is a valid import-style string, return a regex that
     * won't match anything.
     */
    Pattern importStringToPattern(String s) {
        if (com.sun.tools.javac.processing.JavacProcessingEnvironment.isValidImportString(s)) {
            return com.sun.tools.javac.processing.JavacProcessingEnvironment.validImportStringToPattern(s);
        } else {
            Bark bark = Bark.instance(context);
            bark.aptWarning("MalformedSupportedString", s);
            return com.sun.tools.javac.processing.JavacProcessingEnvironment.noMatches;
        }
    }
}
